src/experiment.ts

```diff
@@ -0,0 +1,127 @@
+import { Identify } from 'amplitude-js';
+import { Client } from './client';
+
+const LOCAL_STORAGE_EXPERIMENTS_PREFIX = '__analytics_experiments_';
+
+/** Experiment interface allows getting a variation for a specified user. */
+export interface Experiment<Variation extends string> {
+	name: string;
+
+	engage(userId: string): Variation;
+}
+
+interface VariationData<Variation> {
+	variation: Variation;
+	fraction: number;
+	margin: number;
+}
+
+/** LocalExperiment allows describing an Experiment implemented with local storage. */
+export class LocalExperiment<Variation extends string>
+	implements Experiment<Variation> {
+	private readonly data: Array<VariationData<Variation>> = [];
+	private defaultVariation?: Variation;
+
+	constructor(
+		public readonly name: string,
+		private readonly analytics?: Client,
+	) {}
+
+	private static checkPercent(f: number) {
+		if (isNaN(f) || f < 0 || f > 100) {
+			throw new Error(
+				'Variation fraction must be defines as a percent value between 0 and 100',
+			);
+		}
+	}
+
+	private checkDuplicates(variation: Variation) {
+		const present = this.data.find(d => d.variation === variation);
+		if (present != null) {
+			throw new Error(
+				`Variation [${present.variation} ${present.fraction}%] already exists in experiment ${this.name}.`,
+			);
+		}
+	}
+
+	private static dataString(data: Array<VariationData<any>>): string {
+		return data.map(d => `variation ${d.variation}: ${d.fraction}%`).join(', ');
+	}
+
+	define(variation: Variation, fraction: number): LocalExperiment<Variation> {
+		LocalExperiment.checkPercent(fraction);
+		this.checkDuplicates(variation);
+
+		let previousMargin = 0;
+		if (this.data.length > 0) {
+			previousMargin = this.data[this.data.length - 1].margin;
+		}
+		const margin = previousMargin + fraction;
```

Comment:
The `margin` handling is quite strange, why do we need to keep it as part of each experiment variation? Let's make it a field of the `LocalExperiment` class that gets incremented, or even better, remove it completely. Also, why would we even want to support experiments that don't add up to 100%? What do we show to the rest if all experiments don't add up to 100%?

What might be better is to accept an array of `VariationData` in the `define` method, which will include the default variation as well, and check if the sum of those adds up to 100%. After all, `default` is not a special case, so there is no need to treat it as such.